substrate/primitives/merkle-mountain-range/src/lib.rs

```diff
@@ -361,6 +361,29 @@ pub struct Proof<Hash> {
 	pub items: Vec<Hash>,
 }
 
+/// An MMR proof data for a group of leaves.
+#[derive(codec::Encode, codec::Decode, RuntimeDebug, Clone, PartialEq, Eq, TypeInfo)]
+pub struct NodeProof<Hash> {
+	/// The indices of the leaves the proof is for.
+	pub leaf_indices: Vec<LeafIndex>,
+	/// Number of leaves in MMR, when the proof was generated.
+	pub leaf_count: NodeIndex,
+	/// Proof elements (positions and hashes of siblings of inner nodes on the path to the
+	/// node(s)).
+	pub items: Vec<(u64, Hash)>,
+}
+
+/// An MMR ancestry proof for a prior mmr root.
+#[derive(codec::Encode, codec::Decode, RuntimeDebug, Clone, PartialEq, Eq, TypeInfo)]
+pub struct AncestryProof<Hash> {
```

Comment:
@Lederstrumpf @acatangiu 

Dumb question: 

In order to prove an equivocation at a historical block, can't we just use leaf proofs ? I'm thinking of something like the following:
- for example let's say that we are at block 1 mil and we want to prove an equivocation for a commitment for block 1000 
- we generate a leaf proof for block 1001 at block 1 mil on the client and we add it to the equivocation proof. This should contain a leaf proof plus the leaf data:
    ```
    pub struct MmrLeaf<BlockNumber, Hash, MerkleRoot, ExtraData> {
        /// Version of the leaf format.
        ///
        /// Can be used to enable future format migrations and compatibility.
        /// See [`MmrLeafVersion`] documentation for details.
        pub version: MmrLeafVersion,
        /// Current block parent number and hash.
        pub parent_number_and_hash: (BlockNumber, Hash),
        /// A merkle root of the next BEEFY authority set.
        pub beefy_next_authority_set: BeefyNextAuthoritySet<MerkleRoot>,
        /// Arbitrary extra leaf data to be used by downstream pallets to include custom data in the
        /// [`MmrLeaf`]
        pub leaf_extra: ExtraData,
    }
    ```
- we get the header for block 1000 on the client and we add it to the equivocation proof
- in the runtime, we:
    - check `leaf_1001.parent_number_and_hash.1 == header_1000.hash()`
    - verify the proof for leaf 1001
    - check `extracted mmr payload from header_1000 != equivocated_commitment.payload`

Am I missing something ? Do we actually need ancestry proofs ?

Reply:
The disadvantage of this approach is that from a MMR point of view it is less generic